sdks/java/fn-execution/src/test/java/org/apache/beam/sdk/fn/data/BeamFnDataTimeBasedBufferingOutboundObserverTest.java

```diff
@@ -0,0 +1,139 @@
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Tests for {@link BeamFnDataTimeBasedBufferingOutboundObserver}. */
+@RunWith(JUnit4.class)
+public class BeamFnDataTimeBasedBufferingOutboundObserverTest {
+  private static final LogicalEndpoint OUTPUT_LOCATION = LogicalEndpoint.of("777L", "555L");
+  private static final Coder<WindowedValue<byte[]>> CODER =
+      LengthPrefixCoder.of(WindowedValue.getValueOnlyCoder(ByteArrayCoder.of()));
+
+  @Test
+  public void testConfiguredTimeLimit() throws Exception {
+    Collection<Elements> values = new ArrayList<>();
+    PipelineOptions options = PipelineOptionsFactory.create();
+    options
+        .as(ExperimentalOptions.class)
+        .setExperiments(Arrays.asList("beam_fn_api_data_buffer_time_limit=1"));
+    final CountDownLatch waitForFlush = new CountDownLatch(1);
+    CloseableFnDataReceiver<WindowedValue<byte[]>> consumer =
+        BeamFnDataBufferingOutboundObserver.forLocation(
+            options,
+            OUTPUT_LOCATION,
+            CODER,
+            TestStreams.withOnNext(
+                    (Consumer<Elements>)
+                        e -> {
+                          values.add(e);
+                          waitForFlush.countDown();
+                        })
+                .build());
+
+    // Test that it emits when time passed the time limit
+    consumer.accept(valueInGlobalWindow(new byte[1]));
+    waitForFlush.await(); // wait the flush thread to flush the buffer
+    assertEquals(messageWithData(new byte[1]), Iterables.get(values, 0));
+  }
+
+  @Test
+  public void testConfiguredTimeLimitExceptionPropagation() throws Exception {
+    PipelineOptions options = PipelineOptionsFactory.create();
+    options
+        .as(ExperimentalOptions.class)
+        .setExperiments(Arrays.asList("beam_fn_api_data_buffer_time_limit=1"));
+    BeamFnDataTimeBasedBufferingOutboundObserver<WindowedValue<byte[]>> consumer =
+        (BeamFnDataTimeBasedBufferingOutboundObserver<WindowedValue<byte[]>>)
+            BeamFnDataBufferingOutboundObserver.forLocation(
+                options,
+                OUTPUT_LOCATION,
+                CODER,
+                TestStreams.withOnNext(
+                        (Consumer<Elements>)
+                            e -> {
+                              throw new RuntimeException("");
+                            })
+                    .build());
+
+    // Test that it emits when time passed the time limit
+    consumer.accept(valueInGlobalWindow(new byte[1]));
+    // wait the flush thread to flush the buffer
+    while (consumer.flushException.get() == null) {
+      Thread.sleep(1);
+    }
+    try {
+      // Test that the exception caught in the flush thread is propagate to
+      // the main thread when processing the next element
+      consumer.accept(valueInGlobalWindow(new byte[1]));
+      fail();
+    } catch (IOException e) {
+      // expected
+    }
+
+    consumer =
```

Comment:
It seems like your duplicating the above portion of the test here unless I'm not seeing what the small difference is.

Reply:
The first test case verifies that the exception is propagated when processing the next element. The second test case verifies that the exception is propagated when closing.